src/wp-admin/includes/class-wp-plugin-dependencies.php

```diff
@@ -0,0 +1,688 @@
+<?php
+/**
+ * Dependencies manager for plugins.
+ *
+ * @package dependencies-manager.
+ * @since 5.9.0
+ */
+
+/**
+ * Plugins dependencies manager.
+ *
+ * @since 5.9.0
+ */
+class WP_Plugin_Dependencies {
+
+	/**
+	 * The database option where we store the array of plugins that should be active
+	 * but are not due to unmet dependencies.
+	 *
+	 * @since 5.9.0
+	 * @access private
+	 *
+	 * @var string
+	 */
+	const PENDING_PLUGIN_ACTIVATIONS_OPTION = 'pending_plugin_activations';
+
+	/**
+	 * Installed plugins.
+	 *
+	 * @since 5.9.0
+	 * @access private
+	 *
+	 * @var array[]
+	 */
+	private $installed_plugins;
+
+	/**
+	 * An array of admin notices to show.
+	 *
+	 * @since 5.9.0
+	 * @access private
+	 *
+	 * @var array<int, array>
+	 */
+	private $notices = array();
+
+	/**
+	 * Array of parents for dependencies.
+	 *
+	 * @since 5.9.0
+	 * @access private
+	 *
+	 * @var array<string, array>
+	 */
+	private $dependencies_parents = array();
+
+	/**
+	 * An array of plugin dependencies.
+	 *
+	 * @since 5.9.0
+	 * @access private
+	 *
+	 * @var array<string, array>
+	 */
+	private $plugin_dependencies = array();
+
+	/**
+	 * An array of plugins participating in a circular dependencies loop.
+	 *
+	 * @since 5.9.0
+	 * @access private
+	 *
+	 * @var array<string, bool>
+	 */
+	private $circular_dependencies = array();
+
+	/**
+	 * Constructor.
+	 *
+	 * Add hooks.
+	 *
+	 * @since 5.9.0
+	 */
+	public function __construct() {
+
+		// Early exit if DISALLOW_FILE_MODS or DISALLOW_PLUGIN_DEPENDENCIES is enabled.
+		if (
+			( defined( 'DISALLOW_FILE_MODS' ) && DISALLOW_FILE_MODS ) ||
+			( defined( 'DISALLOW_PLUGIN_DEPENDENCIES' ) && DISALLOW_PLUGIN_DEPENDENCIES )
+		) {
+			return;
+		}
+
+		// Get an array of installed plugins and set it in the object's $installed_plugins prop.
+		$this->get_plugins();
+
+		// Add a hook to allow canceling an activation request.
+		$this->cancel_activation_request();
+
+		// Go through installed plugins and process their dependencies.
+		$this->loop_installed_plugins();
+
+		// Add the admin notices.
+		add_action( 'admin_notices', array( $this, 'admin_notices' ) );
+
+		// Add extra info below plugins that are dependencies.
+		add_action( 'after_plugin_row', array( $this, 'after_plugin_row' ), 10, 2 );
+
+		// Filter available plugin actions.
+		add_filter( 'plugin_action_links', array( $this, 'plugin_action_links' ), 10, 2 );
+
+		add_filter( 'plugin_checkbox_disabled', array( $this, 'disable_checkbox_for_dependency' ), 10, 2 );
+	}
+
+	/**
+	 * Get an array of installed plugins and set it in the object's $installed_plugins prop.
+	 *
+	 * @since 5.9.0
+	 * @access private
+	 *
+	 * @return array[]
+	 */
+	private function get_plugins() {
```

Comment:
A lot of `private` methods in here that I am not sure need to be. It prevents sub-classes from accessing the function which seems problematic if a plugin wishes to extend the functionality somehow. At most they can be `protected`.